pnl_calculation.py

```diff
@@ -0,0 +1,56 @@
+"""
+A file to build cost objects.
+"""
+
+from collections import defaultdict
+
+class CurrencyPnl:
+    """
+    Cost function of currency trading.
+    """
+    def __init__(self, base_ins, instruments):
+        self.base_ins = base_ins
+        self.instruments = instruments
+        self.quantity = defaultdict(int)
+        self.value = defaultdict(int)
+        self.value[base_ins] = 1
+        self.quantity[base_ins] = 1
+
+    def __call__(self, portfolio, price, liquidation):
+        if liquidation:
+            for instrument in self.instruments:
+                portfolio[-1][instrument] = (
+                    0 if instrument is not self.base_ins else 1)
+        return self.calculate(portfolio[-1], price)
+
+    def calculate(self, today, price):
+        """
+        Calculate pnl and produce cost.
+
+        We assume that our overall investment size is always 1.
+        Thus, performance of alphas is more readable because it stands for
+        ratio as well.
+        Please check wiki page for more details.
+        """
+        pnl = {instrument: 0 for instrument in self.instruments}
+        cost = {instrument: 0 for instrument in self.instruments}
+        for instrument in self.instruments:
+            if (price[instrument][-1].buy is not None and
+                    price[instrument][-1].sell is not None):
+                position = today.get(instrument, 0)
+                average_price = (
```

Comment:
average_price is always last middle_price.
Can you cache last middle_price to avoid too many divide operations here?